lib/cache/inventory/inventory_cache_test.go

```diff
@@ -807,6 +808,841 @@ func TestInventoryCacheRateLimiting(t *testing.T) {
+
+		// Test sorting by version while filtering for only instances
+		resp, err = inventoryCache.ListUnifiedInstances(ctx, &inventoryv1.ListUnifiedInstancesRequest{
+			PageSize: 100,
+			Sort:     inventoryv1.UnifiedInstanceSort_UNIFIED_INSTANCE_SORT_VERSION,
+			Order:    inventoryv1.SortOrder_SORT_ORDER_ASCENDING,
+			Filter: &inventoryv1.ListUnifiedInstancesFilter{
+				InstanceTypes: []inventoryv1.InstanceType{inventoryv1.InstanceType_INSTANCE_TYPE_INSTANCE},
+			},
+		})
+		require.NoError(t, err)
+		require.Len(t, resp.Items, 7)
+
+		expectedVersions = []string{"v18.0.0", "18.1.0", "18.2.5", "19.0.0", "19.1.0-alpha", "19.1.0-beta.1", "not-a-version"}
+		for i, item := range resp.Items {
+			require.Equal(t, expectedVersions[i], item.GetInstance().Spec.Version)
+		}
+
+		// Test sorting by version while filtering for only bot instances
+		resp, err = inventoryCache.ListUnifiedInstances(ctx, &inventoryv1.ListUnifiedInstancesRequest{
+			PageSize: 100,
+			Sort:     inventoryv1.UnifiedInstanceSort_UNIFIED_INSTANCE_SORT_VERSION,
+			Order:    inventoryv1.SortOrder_SORT_ORDER_ASCENDING,
+			Filter: &inventoryv1.ListUnifiedInstancesFilter{
+				InstanceTypes: []inventoryv1.InstanceType{inventoryv1.InstanceType_INSTANCE_TYPE_BOT_INSTANCE},
+			},
+		})
+		require.NoError(t, err)
+		require.Len(t, resp.Items, 3)
+
+		expectedBotVersions := []string{"18.0.0", "18.1.5", "19.2.0"}
+		for i, item := range resp.Items {
+			require.Equal(t, expectedBotVersions[i], item.GetBotInstance().Status.LatestHeartbeats[0].Version)
+		}
+	})
+}
+
+// TestSemverSorting tests that semver ordering works properly
+func TestSemverSorting(t *testing.T) {
+	// Expected ordering
+	versions := []struct {
+		version string
+		desc    string
+	}{
+		{"1.0.0-1", "prerelease numeric 1"},
+		{"1.0.0-2", "prerelease numeric 2"},
+		{"1.0.0-11", "prerelease numeric 11"},
+		{"1.0.0-alpha", "prerelease alpha"},
+		{"1.0.0-alpha.1", "prerelease alpha.1"},
+		{"1.0.0-beta", "prerelease beta"},
+		{"1.0.0", "release"},
+		{"1.0.1", "patch increment"},
+		{"1.1.0", "minor version increment"},
+		{"2.0.0", "major version increment"},
+		{"invalid", "invalid version"},
+	}
+
+	// Create the instances
+	instances := make([]*inventoryInstance, len(versions))
+	for i, v := range versions {
+		instances[i] = &inventoryInstance{
+			instance: &types.InstanceV1{
+				ResourceHeader: types.ResourceHeader{
+					Metadata: types.Metadata{
+						Name: fmt.Sprintf("instance-%d", i),
+					},
+				},
+				Spec: types.InstanceSpecV1{
+					Hostname: fmt.Sprintf("host-%d", i),
+					Version:  v.version,
+				},
+			},
+		}
+	}
+
+	// Sort instances by version key
+	sorted := make([]*inventoryInstance, len(instances))
+	copy(sorted, instances)
+	slices.SortFunc(sorted, func(a, b *inventoryInstance) int {
+		keyA := a.getVersionKey()
+		keyB := b.getVersionKey()
+		if keyA < keyB {
+			return -1
+		} else if keyA > keyB {
+			return 1
+		}
+		return 0
+	})
```

Comment:
Alternative way to get a sorted version without first making a temporary (unsorted) copy.

```suggestion
	sorted := slices.SortedFunc(slices.Values(instances), func(a, b *inventoryInstance) int {
		keyA := a.getVersionKey()
		keyB := b.getVersionKey()
		if keyA < keyB {
			return -1
		} else if keyA > keyB {
			return 1
		}
		return 0
	})
```